File: java/BTree.java
public class BTree {
    private static class Node {
        int item;
        Node left;
        Node right;
    }

    public void preOrder(Node root) {
        if (root == null)
            return;

        System.out.printf("%s ", root.item);
        preOrder(root.left);
        preOrder(root.right);
    }

    public static void main(String[] args) {
        // rest of code...
    }
}

File: src/day2/btree.rs
pub struct Node {
    pub item: usize,
    pub left: Option<Box<Node>>,
    pub right: Option<Box<Node>>,
}

pub fn pre_order(root: Node) {
    po(Some(Box::new(root)));
    println!()
}

fn po<T: AsRef<Node>>(node: Option<T>) {
    if let Some(n) = node {
        let n: &Node = <T as AsRef<Node>>::as_ref(&n);
        print!("{} ", &n.item);
        po(n.left.as_ref());
        po(n.right.as_ref());
    }
}
